ethcore/src/engines/clique/tests.rs

```diff
@@ -0,0 +1,776 @@
+		let block = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Add), &prev_header,
+												Some(tester.signers[&'F'].address()), *sign).unwrap();
+		prev_header = block.header().clone();
+	}
+
+	let tags = tester.into_tags(&tester.clique_signers(&prev_header.hash()));
+	assert_eq!(&tags, &['A', 'B', 'C', 'D', 'E', 'F'], "F should have been added");
+
+	// Vote against `F` from [`D`, `E`, `B`, `C`]
+	for sign in SIGNER_TAGS.iter().skip(3).chain(SIGNER_TAGS.iter().skip(1).take(2)) {
+		let block = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Remove), &prev_header,
+												Some(tester.signers[&'F'].address()), *sign).unwrap();
+		prev_header = block.header().clone();
+	}
+
+	let tags = tester.into_tags(&tester.clique_signers(&prev_header.hash()));
+	assert_eq!(&tags, &['A', 'B', 'C', 'D', 'E'], "F should have been removed");
+
+	// Vote for `F` from [`D`, `E`]
+	for sign in SIGNER_TAGS.iter().skip(3).take(2) {
+		let block = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Add), &prev_header,
+												Some(tester.signers[&'F'].address()), *sign).unwrap();
+		prev_header = block.header().clone();
+	}
+
+	// Vote against `A` from [`B`, `C`, `D`]
+	for sign in SIGNER_TAGS.iter().skip(1).take(3) {
+		let block = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Remove), &prev_header,
+												Some(tester.signers[&'A'].address()), *sign).unwrap();
+		prev_header = block.header().clone();
+	}
+
+	let tags = tester.into_tags(&tester.clique_signers(&prev_header.hash()));
+	assert_eq!(&tags, &['B', 'C', 'D', 'E'], "A should have been removed");
+
+	let block = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Add), &prev_header,
+											Some(tester.signers[&'F'].address()), 'B').unwrap();
+
+	let tags = tester.into_tags(&tester.clique_signers(&block.header().hash()));
+	assert_eq!(&tags, &['B', 'C', 'D', 'E', 'F'], "F should have been added again");
+}
+
+#[test]
+fn epoch_transition_reset_all_votes() {
+	let tester = CliqueTester::with(3, 1, vec!['A', 'B']);
+
+	let block = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Add), &tester.genesis,
+											Some(tester.signers[&'C'].address()), 'A').unwrap();
+
+	let block = tester.new_block_and_import(CliqueBlockType::Empty, &block.header(), None, 'B').unwrap();
+	let block = tester.new_block_and_import(CliqueBlockType::Checkpoint, &block.header(), None, 'A').unwrap();
+
+	let block = tester.new_block_and_import(CliqueBlockType::Vote(VoteType::Add), &block.header(),
+											Some(tester.signers[&'C'].address()), 'B').unwrap();
+
+	let tags = tester.into_tags(&tester.clique_signers(&block.header().hash()));
+	assert_eq!(&tags, &['A', 'B'], "Votes should have been reset after checkpoint");
+}
+
+#[test]
+fn unauthorized_signer_should_not_be_able_to_sign_block() {
+	let tester = CliqueTester::with(3, 1, vec!['A']);
+	assert!(tester.new_block_and_import(CliqueBlockType::Empty, &tester.genesis, None, 'B').is_err());
```

Comment:
In the EIP consensus tests there are different errors, I think we should capture those in some enum and check that we're getting the right error

Reply:
That one is quite involved because then I/we should replace all `Boxed errs` with enums. I prefer to keep out of this PR because it already quite large!